signedauth/example/mgr.go

```diff
@@ -0,0 +1,101 @@
+package main
+
+import (
+	"crypto/md5"
+	"crypto/sha1"
+	"encoding/hex"
+	"errors"
+	"github.com/ChristopherRabotin/gin-contrib/signedauth"
+	"hash"
+	"io/ioutil"
+	"net/http"
+	"time"
+)
+
+// StrictSHA1Manager is an example definition of an AuthKeyManager struct.
+type StrictSHA1Manager struct {
+	prefix string
+	secret string
+	key    string
+	value  interface{}
+}
+
+// AuthHeaderPrefix returns the prefix used in the initialization.
+func (mgr StrictSHA1Manager) AuthHeaderPrefix() string {
+	return mgr.prefix
+}
+
+// SecretKey returns the secret key from the provided access key.
+// Here should reside additional verifications on the header, or other parts of the request, if needed.
+func (mgr StrictSHA1Manager) SecretKey(access string, req *http.Request) (string, *signedauth.Error) {
+	if req.ContentLength != 0 && req.Body == nil {
+		// Not sure whether net/http or Gin handles these kinds of fun situations.
+		return "", &signedauth.Error{400, errors.New("Received a forged packet.")}
+	}
+	// Grabbing the date and making sure it's in the correct format and is within fifteen minutes.
+	if dateHeader := req.Header.Get("Date"); dateHeader == "" {
+		return "", &signedauth.Error{406, errors.New("No Date header provided.")}
+	} else {
+		date, derr := time.Parse("2006-01-02T15:04:05.000Z", dateHeader)
+		if derr != nil {
+			return "", &signedauth.Error{408, errors.New("Could not parse date.")}
+		} else if time.Since(date) > time.Minute*15 {
+			return "", &signedauth.Error{410, errors.New("Request is too old.")}
+		}
+	}
+	// The headers look good, let's check the access key.
+	// If the reading the access key requires any kind of IO (database, or file reading, etc.)
+	// it's quite good to only verify if that access key is valid once all the checks are done.
+	if access == "my_access_key" {
+		return mgr.secret, nil
+	}
+	return "", &signedauth.Error{418, errors.New("You are a teapot.")}
+}
+
+// ContextKey returns the key which will store the return from ContextValue() in Gin's context.
+func (mgr StrictSHA1Manager) ContextKey() string {
+	return mgr.key
+}
+
+// ContextValue returns the value to store in Gin's context at ContextKey().
+func (mgr StrictSHA1Manager) ContextValue(access string) interface{} {
+	if access == "my_access_key" {
```

Comment:
The `ContextValue` function allows to set a given value into the context provided an access string, and is only called once the signature was proven as valid. Hence, this function takes, as input, some uniquely user-identifiable string and sets a given value in the context for subsequent calls in the context. For example if user with ID 5 has access key `user_id_5_access` with a given secret key, and passed all the validation steps for the request, then this function will be called with `user_id_5_access` as its only parameter, and can set `5` in the `user` variable of the context. This is an example of course because everything is configurable in this plugin, as explained in the README.